### PR TITLE
[WIP] Port of nav2_costmap_2d tests

### PR DESCRIPTION
**Pull request by bpwilcox**
---

## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | #121  |
| Primary OS tested on | Ubuntu 16.04 |
---

## Description of contribution
- Adds port of costmap_2d tests including 
  - footprint_tests
  - static_tests (the code in this test is commented out in the original navigation stack)
  - inflation_tests
  - obstacle tests
  - costmap_tester (ported but not enabled until rosbag playback is available)
---

## Future work that may be required
The 'module_tests.cpp' is not being built in the original navstack costmap_2d, so we should consider porting or deleting after investigating what it helps to test. The test of the 'costmap_tester.cpp' seems to have been launched with a rosbag and the 'simple_driving_test.xml'. Rosbags might not be available now, but we should investigate testing the `Costmap2DROS` in a full system test

---

### CODE REVIEW

File: nav2_costmap_2d/test/integration/costmap_tester.cpp [hunk 1]
```diff
@@ -52,13 +55,17 @@ class CostmapTester : public testing::Test {
...
   private:
-    nav2_costmap_2d::Costmap2DROS costmap_ros_;
+    std::shared_ptr<nav2_costmap_2d::Costmap2DROS> costmap_ros_;
 };
 
-CostmapTester::CostmapTester(tf2_ros::Buffer& tf): costmap_ros_("test_costmap", tf){}
+CostmapTester::CostmapTester(tf2_ros::Buffer& tf)
+{
+  costmap_ros_ = std::make_shared<nav2_costmap_2d::Costmap2DROS>("test_costmap", tf);
+  //rclcpp::spin(costmap_ros_);
```
Comment (mjeronimo) [inline nav2_costmap_2d/test/integration/costmap_tester.cpp]: Definitely shouldn't spin in the constructor. Should just delete the commented-out line.

File: nav2_costmap_2d/test/integration/costmap_tester.cpp [hunk 2]
```diff
@@ -103,9 +110,9 @@ void CostmapTester::compareCells(nav2_costmap_2d::Costmap2D& costmap, unsigned i
     double furthest_valid_distance = 0; // ################costmap.cell_inscribed_radius_ + cell_distance + 1;
```
Comment (mjeronimo) [inline nav2_costmap_2d/test/integration/costmap_tester.cpp]: Is this commented-out code still needed? Should delete.

File: nav2_costmap_2d/test/integration/costmap_tester.cpp [hunk 3]
```diff
@@ -103,9 +110,9 @@ void CostmapTester::compareCells(nav2_costmap_2d::Costmap2D& costmap, unsigned i
...
     unsigned char expected_lowest_cost = 0; // ################costmap.computeCost(furthest_valid_distance);
     if(neighbor_cost < expected_lowest_cost){
-      ROS_ERROR("Cell cost (%d, %d): %d, neighbor cost (%d, %d): %d, expected lowest cost: %d, cell distance: %.2f, furthest valid distance: %.2f",
+      RCLCPP_ERROR(rclcpp::get_logger("costmap_tester"), "Cell cost (%d, %d): %d, neighbor cost (%d, %d): %d, expected lowest cost: %d, cell distance: %.2f, furthest valid distance: %.2f",
```
Comment (mjeronimo) [inline nav2_costmap_2d/test/integration/costmap_tester.cpp]: Need to run the linting tools. Lines too long.

File: nav2_costmap_2d/test/integration/footprint_tests.cpp [hunk 4]
```diff
@@ -143,40 +169,25 @@ TEST( Costmap2DROS, footprint_from_same_level_param )
...
-}
-
 int main(int argc, char** argv)
 {
-  ros::init(argc, argv, "footprint_tests_node");
+  rclcpp::init(argc, argv);
+  auto node = rclcpp::Node::make_shared("footprint_tests");
 
-  tf_ = new tf2_ros::Buffer( ros::Duration( 10 ));
+  tf_ = new tf2_ros::Buffer(node->get_clock(), tf2::durationFromSec(10));
```
Comment (crdelsey) [inline nav2_costmap_2d/test/integration/footprint_tests.cpp]: You could leave out the durationFromSec(10) bit. That is the default value for the Buffer constructor.

File: nav2_costmap_2d/test/integration/costmap_tester.cpp (hunk 3 above)
Comment (bpwilcox) [reply nav2_costmap_2d/test/integration/costmap_tester.cpp]: I'd like to run linting tools on all the nav2_costmap_2d code in one PR. I made issue #396 to track

**Comment by mkhansen-intel** [comment]
Can we merge this now?

**Comment by bpwilcox** [comment]
> Can we merge this now?

There still exists the bug where there is a segfault or "double free or corruption" error at the end of the launch test, causing the tests to fail even though apparently the gtests technically pass. I'm digging into this issue.

**Comment by mkhansen-intel** [comment]
Changed title to [WIP] so when it's ready for final review you can change it back and we'll know to look at it again.

**Comment by SteveMacenski** [comment]
Please consider waiting on merging any test/dependencies related code until after the first sync.

**Comment by bpwilcox** [comment]
I commented out the footprint and inflation tests that, though passing gtests, appear to have bugs (#508) upon breakdown/shutdown after the test.

**Pull request merged**